File: tests/algorithms/transpiler/FS/sorts/patience_sort.fs
// Generated 2025-08-11 17:23 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let rec _str v =
    let s = sprintf "%A" v
    let s = if s.EndsWith(".0") then s.Substring(0, s.Length - 2) else s
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec bisect_left (stacks: int array array) (value: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable stacks = stacks
    let mutable value = value
    try
        let mutable low: int = 0
        let mutable high: int = Seq.length (stacks)
        while low < high do
            let mid: int = _floordiv (low + high) 2
            let stack: int array = _idx stacks (int mid)
            let top_idx: int = (Seq.length (stack)) - 1
            let top: int = _idx stack (int top_idx)
            if top < value then
                low <- mid + 1
            else
                high <- mid
        __ret <- low
        raise Return
        __ret
    with
        | Return -> __ret
and reverse_list (src: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable src = src
    try
        let mutable res: int array = Array.empty<int>
        let mutable i: int = (Seq.length (src)) - 1
        while i >= 0 do
            res <- Array.append res [|(_idx src (int i))|]
            i <- i - 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and patience_sort (collection: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable collection = collection
    try
        let mutable stacks: int array array = Array.empty<int array>
        let mutable i: int = 0
        while i < (Seq.length (collection)) do
            let element: int = _idx collection (int i)
            let idx: int = bisect_left (stacks) (element)
            if idx <> (Seq.length (stacks)) then
                let stack: int array = _idx stacks (int idx)
                stacks.[int idx] <- Array.append stack [|element|]
            else
                let mutable new_stack: int array = unbox<int array> [|element|]
                stacks <- Array.append stacks [|new_stack|]
            i <- i + 1
        i <- 0
        while i < (Seq.length (stacks)) do
            stacks.[int i] <- reverse_list (_idx stacks (int i))
            i <- i + 1
        let mutable indices: int array = Array.empty<int>
        i <- 0
        while i < (Seq.length (stacks)) do
            indices <- Array.append indices [|0|]
            i <- i + 1
        let mutable total: int = 0
        i <- 0
        while i < (Seq.length (stacks)) do
            total <- total + (Seq.length (_idx stacks (int i)))
            i <- i + 1
        let mutable result: int array = Array.empty<int>
        let mutable count: int = 0
        while count < total do
            let mutable min_val: int = 0
            let mutable min_stack: int = -1
            let mutable j: int = 0
            while j < (Seq.length (stacks)) do
                let idx: int = _idx indices (int j)
                if idx < (Seq.length (_idx stacks (int j))) then
                    let ``val``: int = _idx (_idx stacks (int j)) (int idx)
                    if min_stack < 0 then
                        min_val <- ``val``
                        min_stack <- j
                    else
                        if ``val`` < min_val then
                            min_val <- ``val``
                            min_stack <- j
                j <- j + 1
            result <- Array.append result [|min_val|]
            indices.[int min_stack] <- (_idx indices (int min_stack)) + 1
            count <- count + 1
        i <- 0
        while i < (Seq.length (result)) do
            collection.[int i] <- _idx result (int i)
            i <- i + 1
        __ret <- collection
        raise Return
        __ret
    with
        | Return -> __ret
printfn "%s" (_str (patience_sort (unbox<int array> [|1; 9; 5; 21; 17; 6|])))
printfn "%s" (_str (patience_sort (Array.empty<int>)))
printfn "%s" (_str (patience_sort (unbox<int array> [|-3; -17; -48|])))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/sorts/patience_sort.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if ($b === 0 || $b === '0') {
        throw new DivisionByZeroError();
    }
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
function _iadd($a, $b) {
    if (function_exists('bcadd')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcadd($sa, $sb, 0);
    }
    return $a + $b;
}
function _isub($a, $b) {
    if (function_exists('bcsub')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcsub($sa, $sb, 0);
    }
    return $a - $b;
}
function _imul($a, $b) {
    if (function_exists('bcmul')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcmul($sa, $sb, 0);
    }
    return $a * $b;
}
function _idiv($a, $b) {
    return _intdiv($a, $b);
}
function _imod($a, $b) {
    if (function_exists('bcmod')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcmod($sa, $sb));
    }
    return $a % $b;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function bisect_left($stacks, $value) {
  $low = 0;
  $high = count($stacks);
  while ($low < $high) {
  $mid = _intdiv((_iadd($low, $high)), 2);
  $stack = $stacks[$mid];
  $top_idx = _isub(count($stack), 1);
  $top = $stack[$top_idx];
  if ($top < $value) {
  $low = _iadd($mid, 1);
} else {
  $high = $mid;
}
};
  return $low;
};
  function reverse_list($src) {
  $res = [];
  $i = _isub(count($src), 1);
  while ($i >= 0) {
  $res = _append($res, $src[$i]);
  $i = _isub($i, 1);
};
  return $res;
};
  function patience_sort(&$collection) {
  $stacks = [];
  $i = 0;
  while ($i < count($collection)) {
  $element = $collection[$i];
  $idx = bisect_left($stacks, $element);
  if ($idx != count($stacks)) {
  $stack = $stacks[$idx];
  $stacks[$idx] = _append($stack, $element);
} else {
  $new_stack = [$element];
  $stacks = _append($stacks, $new_stack);
}
  $i = _iadd($i, 1);
};
  $i = 0;
  while ($i < count($stacks)) {
  $stacks[$i] = reverse_list($stacks[$i]);
  $i = _iadd($i, 1);
};
  $indices = [];
  $i = 0;
  while ($i < count($stacks)) {
  $indices = _append($indices, 0);
  $i = _iadd($i, 1);
};
  $total = 0;
  $i = 0;
  while ($i < count($stacks)) {
  $total = _iadd($total, count($stacks[$i]));
  $i = _iadd($i, 1);
};
  $result = [];
  $count = 0;
  while ($count < $total) {
  $min_val = 0;
  $min_stack = -1;
  $j = 0;
  while ($j < count($stacks)) {
  $idx = $indices[$j];
  if ($idx < count($stacks[$j])) {
  $val = $stacks[$j][$idx];
  if ($min_stack < 0) {
  $min_val = $val;
  $min_stack = $j;
} else {
  if ($val < $min_val) {
  $min_val = $val;
  $min_stack = $j;
};
};
}
  $j = _iadd($j, 1);
};
  $result = _append($result, $min_val);
  $indices[$min_stack] = _iadd($indices[$min_stack], 1);
  $count = _iadd($count, 1);
};
  $i = 0;
  while ($i < count($result)) {
  $collection[$i] = $result[$i];
  $i = _iadd($i, 1);
};
  return $collection;
};
  echo rtrim(_str(patience_sort([1, 9, 5, 21, 17, 6]))), PHP_EOL;
  echo rtrim(_str(patience_sort([]))), PHP_EOL;
  echo rtrim(_str(patience_sort([-3, -17, -48]))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
